### PR TITLE
Refactor SamlAuthenticatorTests 

### LINKED ISSUE #49742: SamlAuthenticatorTests fail on zulu8 runtime
These started failing only on zulu8 after Novemember 30th : https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-60d,mode:quick,to:now))&_a=(columns:!(branch),index:e58bf320-7efd-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:SamlAuthenticatorTests),sort:!(time,desc))

It would make sense if https://github.com/elastic/elasticsearch/pull/49512 caused this somehow but this fails in 6.8, 7.4, 7.5 7.x and not on master while the opensaml dependencies were updated in master and 7.x only. 

The stack trace is 
```
java.lang.NoClassDefFoundError: Could not initialize class org.jcp.xml.dsig.internal.dom.DOMEnvelopedTransform
	at __randomizedtesting.SeedInfo.seed([13A666290A5F1991:2BB4D77CB77ADE28]:0)
	at org.jcp.xml.dsig.internal.dom.XMLDSigRI$ProviderService.newInstance(XMLDSigRI.java:111)
	at sun.security.jca.GetInstance.getInstance(GetInstance.java:236)
	at javax.xml.crypto.dsig.TransformService.getInstance(TransformService.java:215)
	at org.jcp.xml.dsig.internal.dom.DOMXMLSignatureFactory.newTransform(DOMXMLSignatureFactory.java:314)
	at org.elasticsearch.xpack.security.authc.saml.SamlAuthenticatorTests.signElement(SamlAuthenticatorTests.java:2358)
	at org.elasticsearch.xpack.security.authc.saml.SamlAuthenticatorTests.signDoc(SamlAuthenticatorTests.java:2264)
	at org.elasticsearch.xpack.security.authc.saml.SamlAuthenticatorTests.signDoc(SamlAuthenticatorTests.java:2251)
	at org.elasticsearch.xpack.security.authc.saml.SamlAuthenticatorTests.testIncorrectSubjectConfirmationDataInResponseToIsRejected(SamlAuthenticatorTests.java:1287)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
	at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at java.lang.Thread.run(Thread.java:748)
```

### PR DESCRIPTION
**Pull request by jkakavas**
- Use opensaml utility classes to sign and encrypt responses/assertions/attributes
instead of manually signing XML documents
- Use opensaml to build Response and Assertion objects instead of
parsing XML strings and replacing values.

These changes allow us to run the tests in Azul Zulu 8 and in
FIPS 140 JVMs. Due to FIPS 140 requirements, BouncyCastle FIPS provider 
will block RSA keys that have been used for signing from being used for 
encryption and vice versa. This commit also adds new encryption specific keys 
to be used throughout the tests.

Resolves: #49742

### CODE REVIEW

**Comment by jkakavas** [comment]
@elasticmachine run elasticsearch-ci/docs

**Comment by tvernum** [comment]
This bothers me, and I'd like to talk it through. 

We used to (attempt to) test that we could parse raw XML correctly.
Now we only test that we can parse OpenSAML's XML output, which is the same library that we use to do parsing, so that seems like a case of "well, yes, that is going to work".

However, this isn't the _only_ set of tests we have, so I think my question is: _After this change, how do we know that we can parse SAML that wasn't produced by OpenSAML?_

**Comment by jkakavas** [comment]
>  After this change, how do we know that we can parse SAML that wasn't produced by OpenSAML?

Fair Question. I think that a more valid question would be how do we know that we can parse valid ( according to the XSDs of the specification ) SAML message XML documents ( irrespective to which implementation produced them ). I agree that this change asserts more trust on the correctness of the opensaml implementation, but I believe this is fine. Note, also, that we do parse the XML documents we handle against the schemas as defined by the SAML specification, _before_ we feed them to opensaml to be unmarshalled, so we should catch issues there. 

> We used to (attempt to) test that we could parse raw XML correctly.

In my view, what we used to do is that we had tests which tested that we could parse XML documents that we ourselves had crafted by hand in order to be successfully parsed. Which also, of course will work. 

In that, it makes more sense to me that we verify that we can parse SAML Messages (as XML Documents) that have been created by a library that is common out there and proven ( to the extent possible ) that produces valid SAML message XML Documents , rather than verifying that we can parse documents that are created to be parsed.

> Now we only test that we can parse OpenSAML's XML output,

Parsing is just a subset of the tests though. We test tons of stuff in this class. 

How about I reintroduce a couple of the tests where we parsed the XML document from a string? Would that give us enough certainty that we can handle SAML messages regardless of what library output those ? I can see the value in that we can catch a case where a bug is introduced in an opensaml version that makes it produce invalid saml messages while it also allows it to parse those, and we would miss that otherwise.

**Comment by jkakavas** [comment]
Any further thoughts on this @tvernum ?

**Comment by tvernum** [comment]
> I think that a more valid question would be how do we know that we can parse valid ( according to the XSDs of the specification ) SAML message XML documents ( irrespective to which implementation produced them ). I agree that this change asserts more trust on the correctness of the opensaml implementation, but I believe this is fine. Note

This isn't really where my concern comes from.
It's not  
1. that we might parse something that's invalid according to the schema, _or_
2. that opensaml will produce bad XML

it's that we ought to have a test suite for real world XML from various implementations that we want to be compatible with. Particularly any that we know have caused us pain.

This test was the place where we would construct such examples if we needed them, but it's moving in the opposite direction where it can only be used to parse XML that OpenSAML is capable of generating. That is, if we have a scenario we want to test, we can only do that if we can work out how to get OpenSAML to produce that XML.
 
> Parsing is just a subset of the tests though. We test tons of stuff in this class.

I think that's the problem. This test did lots of stuff & had a bunch of weak areas. But in order to fix one of the issues, we're reducing its scope and not picking up the pieces that we're dropping

> How about I reintroduce a couple of the tests where we parsed the XML document from a string? 

I think what we probably need is a specific test that focuses on raw XML, and if we need to in the future we can add to that test with `testAuthenticateUsingSignedAssertionsFromBadlyImplementedIdP`.

I don't think we want to keep the XML tests around "just because", but I worry that we don't have a good answer for:
> how do we know that we can parse SAML that wasn't produced by OpenSAML?

and we should work towards that.

**Comment by jkakavas** [comment]
I see your points. I misunderstood previous feedback to imply that we should leave the xml construction as - is and my issue with that was that "crafting all xml docs by hand from the same template" doesn't buy us much compared to "creating all xml docs with opensaml"

I'll add a method to allow us to get raw XML from a template ( as we had ) and we can use that for a specific test that would help us answer

> how do we know that we can parse SAML that wasn't produced by OpenSAML?

and in the future when we want to test with messages that opensaml can't produce/handle - as we're now doing in `testIgnoredCommentsInLegitimateResponses` for instance

**Comment by jkakavas** [comment]
@tvernum I did what I talked about in https://github.com/elastic/elasticsearch/pull/51089#issuecomment-579698639, let me know if you think this is enough or we should do further refactoring of the class.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlAuthenticatorTests.java [hunk 1]
```diff
@@ -2198,21 +1175,47 @@ public void testIgnoredCommentsInForgedResponses() throws Exception {
     }
 
     public void testIgnoredCommentsInLegitimateResponses() throws Exception {
+        assumeFalse("Can't run in a FIPS JVM, there is no DOM XMLSignature Factory so we can't sign XML documents", inFipsJvm());
+        assumeFalse("Can't run in Azul Zulu JVM",System.getProperty("java.vendor", "").contains("Azul"));
```
Comment (tvernum) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/saml/SamlAuthenticatorTests.java:1179]:
```suggestion
        assumeFalse("Can't run in Azul Zulu JVM", System.getProperty("java.vendor", "").contains("Azul"));
```

**Review by tvernum — APPROVED** [review]
LGTM

**Pull request merged**